proxy/src/project.rs

```diff
@@ -80,14 +86,154 @@ pub enum Registration {
+    /// `None` when executing the application for real. However, if we want to run an integration
+    /// test we need to tell say where the `git-rad-remote` helper can be found.
+    bin_path: Option<ffi::OsString>,
+}
+
+impl<P, ST> Checkout<P, ST>
+where
+    P: AsRef<path::Path>,
+    ST: Clone,
+{
+    /// Create a new `Checkout` with the mock `Credential::Password` helper.
+    pub fn new<Bin>(project: coco::Project<ST>, path: P, bin_path: Bin) -> Self
+    where
+        Bin: Into<Option<ffi::OsString>>,
+    {
+        Self {
+            // TODO(rudolfs): we'll have to figure out how to pass the secret
+            // key to git in a safe manner. As it is now it could be sniffed
+            // out from the process list while the user is doing a clone.
+            //
+            // How will we get ahold on the secret key here?
+            credential: Credential::Password("radicle-upstream".to_owned()),
+            project,
+            path,
+            bin_path: bin_path.into(),
+        }
+    }
+
+    /// Checkout a working copy of a [`Project`].
+    ///
+    /// NOTE: `RAD_HOME` should be expected to be set if using a custom root for
+    /// [`librad::paths::Paths`]. If it is not set the underlying binary will delegate to the
+    /// `ProjectDirs` setup of the `Paths`.
+    ///
+    /// # Errors
+    ///
+    ///   * We couldn't resolve the executable path.
+    ///   * The checkout process failed.
+    pub fn run(self) -> Result<path::PathBuf, error::Error> {
+        let bin_path = match self.bin_path {
+            Some(path) => Ok(path),
+            None => Self::default_bin_path(),
+        }?;
+
+        // Check if the path provided ends in the 'directory_name' provided. If not we create the
+        // full path to that name.
+        let path = &self.path.as_ref();
+        let project_path = if let Some(destination) = path.components().next_back() {
+            let destination: &std::ffi::OsStr = destination.as_ref();
+            let project_name = self.project.name().to_string();
+            let name: &std::ffi::OsStr = project_name.as_ref();
+            if destination == name {
+                path.to_path_buf()
+            } else {
+                path.join(name)
+            }
+        } else {
+            path.join(&self.project.name().to_string())
+        };
+
+        let mut child_process = Command::new("git")
+            .arg("-c")
+            .arg(self.credential.to_helper())
+            .arg("clone")
+            .arg("-o")
+            .arg("rad")
+            .arg("-b")
+            .arg(self.project.default_branch())
+            .arg(LocalUrl::from(self.project.urn()).to_string())
+            .arg(project_path.as_os_str())
+            .env("PATH", &bin_path)
+            .envs(std::env::vars().filter(|(key, _)| key.starts_with("GIT_TRACE")))
+            .spawn()?;
+
+        // TODO: Capture the error if any and respond
```

Comment:
Do we think this is possible? :thinking: And/or is still valid?

Reply:
Good point. The UI now shows an error notification if something goes wrong in proxy, but the error messaging could be improved.

This is what the user sees:
<img width="397" alt="Screenshot 2020-08-10 at 18 00 38" src="https://user-images.githubusercontent.com/158411/89803659-6d318b80-db33-11ea-98b1-8173f0e6cfa9.png">

And this is what is logged by proxy:
```
Cloning into '/Users/rudolfs/work/radicle-checkout/rx'...
Error: Signer key does not match key used at initialisation

Stack backtrace:
   0: std::backtrace::Backtrace::create
   1: std::backtrace::Backtrace::capture
   2: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   3: radicle_git_helpers::remote_helper::run
   4: git_remote_rad::main
   5: std::rt::lang_start::{{closure}}
   6: std::rt::lang_start_internal
   7: std::rt::lang_start
   8: main
 ERROR warp::filters::body   > request body already taken in previous filter
 ERROR proxy::http::error    > Rejection([MethodNotAllowed, BodyConsumedMultipleTimes, Checkout, MethodNotAllowed])
```

I'd create a follow-up issue for that to not increase the scope here any further, if you're okay with that.

Reply:
I'm perfectly ok with that :heart:

Reply:
https://github.com/radicle-dev/radicle-upstream/issues/767